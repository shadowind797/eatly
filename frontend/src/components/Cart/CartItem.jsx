import {useEffect, useState} from "react";
import api from "../../api.js";

function CartItem({cartItem}){
    const [dish, setDish] = useState({});
    const [quantity, setQuantity] = useState(cartItem.quantity);
    const [deleted, setDeleted] = useState(false);

    useEffect(() => {
        getItems()
    }, []);

    const getItems = () => {
        api
            .get("api/items/", {params: {id: cartItem.item}})
            .then((res) => res.data)
            .then((data) => {
                data.map((item) => {
                    if (item.id === cartItem.item) {
                        setDish(item)
                    }
                })
            })
            .catch((err) => alert(err));
    }

    const deleteItem = () => {
        api.delete("api/items/cart/delete",
            {params: {id: cartItem.id}})
            .then((res) => {
                if (res.status === 202) {}
                else if (res.status === 404) {}
            }).catch((err) => {});
        setDeleted(true)
    }

    const addQuantity = () => {
        api.post("api/items/cart/add", {item: dish.id, quantity: quantity + 1},
                                                    {params: {method: "addQuant"}})
                                                                .then((res) => {
            if (res.status === 201) {} else {
                alert("Failed to create item");
            }
        }).catch((err) => {});
    }

    const removeQuantity = () => {
        if (quantity > 1) {
            api.post("api/items/cart/add", {item: dish.id, quantity: quantity - 1},
                {params: {method: "addQuant"}})
                .then((res) => {
                    if (res.status === 201) {
                    } else {
                        alert("Failed to create item");
                    }
                }).catch((err) => {
            });
        } else {
            deleteItem()
        }
    }

    if (deleted === false) {
        return (
            <div>
                <img src={dish.photo} alt=""/>
                <p>{quantity}</p>
                <button onClick={() => {
                    setQuantity(quantity + 1)
                    addQuantity()
                }}>addItem
                </button>
                <button onClick={() => {
                    setQuantity(quantity - 1)
                    removeQuantity()
                }}>removeItem
                </button>
            </div>
        )
    } else {
        return (<p></p>)
    }
}

export default CartItem