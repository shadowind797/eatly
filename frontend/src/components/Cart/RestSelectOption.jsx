/* eslint-disable react/prop-types */
import { components } from "react-select";

const RestSelectOption = (props) => {
  return (
    <components.Option {...props}>
      <img
        src={`${import.meta.env.VITE_API_URL}/media${props.data.logo}`}
        alt=""
        style={{ width: 20, height: 20, marginRight: 10 }}
      />
      {props.data.label}
    </components.Option>
  );
};

export default RestSelectOption;
