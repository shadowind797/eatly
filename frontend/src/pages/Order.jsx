import Info from "../components/Order/Info.jsx";
import {useEffect, useState} from "react";
import api from "../api.js";
import Map from "../components/Order/Map.jsx";


function Order() {
    const [user, setUser] = useState({});
    const [addPayment, setAddPayment] = useState(false);
    const [paymentMode, setPaymentMode] = useState("Cash");
    const [order, setOrder] = useState({});
    const [address, setAddress] = useState({});

    useEffect(() => {
        getUser()
        getOrder()
    }, []);

    const getUser = () => {
        api
            .get("api/user/")
            .then((res) => res.data)
            .then((data) => {
                data.map((item) => {
                    setUser(item);
                })
            })
            .catch((err) => {
            });
    }

    const getOrder = () => {
        api
            .get("api/order/", {params: {method: "for_complete"}})
            .then((res) => res.data)
            .then((data) => {
                setOrder(data)
            })
            .catch((err) => {
            });
    }


    return (
        <div id="complete-order">
            <Map/>
            <Info order={order} user={user}/>
        </div>
    )
}

export default Order;