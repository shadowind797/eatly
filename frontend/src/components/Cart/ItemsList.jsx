import CartItem from "./CartItem.jsx";
import PropTypes from "prop-types";

ItemsList.propTypes = {
  items: PropTypes.array.isRequired,
  cartItems: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
  rest: PropTypes.object,
};

function ItemsList({ items, cartItems, onChange, rest }) {
  return (
    <div>
      <div id="rest-name">
        <img src={rest ? `${import.meta.env.VITE_API_URL}/media${rest.logo}` : ""} alt="" />
        <h2>
          {rest && rest.name} - <span>{items && items.length} dishes</span>
        </h2>
      </div>
      <div id="cart-items-list" className="container">
        {items.map((item) => (
          <CartItem
            dish={item}
            cartItem={cartItems.find((ci) => ci.item_id === item.id)}
            onChange={onChange}
            key={item.id}
          />
        ))}
      </div>
    </div>
  );
}

export default ItemsList;
