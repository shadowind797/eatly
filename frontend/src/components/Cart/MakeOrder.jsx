import api from "../../api.js";
import { useEffect, useState } from "react";
import PropTypes from "prop-types";
import Select from "react-select";
import { Navigate } from "react-router-dom";
import Map from "./Map.jsx";
import PlaceInput from "./PlaceInput.jsx";
import cpImg from "../../assets/coupon.svg";
import RestSelectOption from "./RestSelectOption.jsx";

MakeOrder.propTypes = {
  subtotal: PropTypes.string.isRequired,
  createOrder: PropTypes.func.isRequired,
  test: PropTypes.bool,
  rests: PropTypes.array.isRequired,
  changeItems: PropTypes.func.isRequired,
  items: PropTypes.array.isRequired,
  updateTotal: PropTypes.func.isRequired,
  updateRest: PropTypes.func.isRequired,
};

function MakeOrder({
  subtotal,
  createOrder,
  test,
  rests,
  changeItems,
  items,
  updateTotal,
  updateRest,
}) {
  const [user, setUser] = useState({});
  const [coupon, setCoupon] = useState("");
  const [couponValue, setCouponValue] = useState(1);
  const [applied, setApplied] = useState(null);
  const [firstName, setFirstName] = useState(user.first_name || "");
  const [discount, setDiscount] = useState(0);
  const [phone, setPhone] = useState(user.phone || "");
  const [restaurant, setRestaurant] = useState(rests[0].id);
  const [restList, setRestList] = useState([]);

  const [addressList, setAddressList] = useState([]);
  const [address, setAddress] = useState(null);
  const [addAddress, setAddAddress] = useState(false);
  const [addressAlreadyExists, setAddressAlreadyExists] = useState(false);

  const [buildingAddress, setBuildingAddress] = useState("");
  const [entrance, setEntrance] = useState("");
  const [floor, setFloor] = useState("");
  const [flat, setFlat] = useState("");

  const [toComplete, setToComplete] = useState(false);
  const [noItems, setNoItems] = useState(false);
  const [noName, setNoName] = useState(false);
  const [noAddress, setNoAddress] = useState(false);
  const [noPhone, setNoPhone] = useState(false);
  const [alreadyExists, setAlreadyExists] = useState(false);
  const [noBuilding, setNoBuilding] = useState(false);
  const [noCoupon, setNoCoupon] = useState(false);
  const [discardCouponBtn, setDiscardCouponBtn] = useState(false);

  const [addressLoading, setAddressLoading] = useState(false);
  const [mapLoading, setMapLoading] = useState(false);
  const [couponLoading, setCouponLoading] = useState(false);
  const [mapSelectMode, setMapSelectMode] = useState(false);
  const [mapsError, setMapsError] = useState("");

  useEffect(() => {
    setDiscount(
      Math.abs(subtotal * 1.1 - subtotal * 1.1 * couponValue).toFixed(2)
    );
  }, [couponValue]);

  useEffect(() => {
    if (restaurant) {
      const oneRestItems = items.filter(
        (item) => item.restaurant_id === restaurant
      );
      updateRest(restaurant);
      changeItems(oneRestItems);
      updateTotal(oneRestItems, restaurant);
    }
  }, [restaurant]);

  useEffect(() => {
    getUser();
    getAddressList();
  }, []);

  useEffect(() => {
    if (rests) {
      let list = [];
      rests.map((item) => {
        list = [
          ...list,
          {
            value: item.id,
            label: item.name,
            logo: item.logo,
          },
        ];
      });
      setRestList(list);
    }
  }, [rests]);

  const changeAddress = (address) => {
    setAddress(address.value);
  };

  const changeRest = (rest) => {
    setRestaurant(rest.value);
  };

  const getTotal = () => {
    if (couponValue === 1) {
      return (subtotal * 1.1).toFixed(2);
    } else {
      return ((subtotal * 1.1).toFixed(2) - discount).toFixed(2);
    }
  };

  const getUser = () => {
    api
      .get("api/user/")
      .then((res) => res.data)
      .then((data) => {
        data.map((item) => {
          setUser(item);
          setFirstName(item.first_name);
          setPhone(item.phone);
        });
      })
      .catch(() => {});
  };

  const checkAddressFormat = (address) => {
    let formattedAddress = address.house_address;
    if (address.entrance.length > 0) {
      formattedAddress += `, ${address.entrance} ent.`;
    }
    if (address.floor.length > 0) {
      formattedAddress += `, ${address.floor} floor`;
    }
    if (address.flat.length > 0) {
      formattedAddress += `, flat ${address.flat}`;
    }
    return formattedAddress;
  };

  const getAddressList = (add) => {
    api
      .get("api/address/")
      .then((res) => res.data)
      .then((data) => {
        let list = [];
        data.map((item) => {
          list = [
            ...list,
            {
              value: item.id,
              label: checkAddressFormat(item),
            },
          ];
          setBuildingAddress("");
          setEntrance("");
          setFloor("");
          setFlat("");
          setAddressList(list);
          setNoBuilding(false);
          setAddressAlreadyExists(false);
          setMapSelectMode(false);
        });
        if (add) {
          setAddAddress(false);
          setAddressLoading(false);
        }
      })
      .catch(() => {});
  };

  const checkCoupon = (e) => {
    e.preventDefault();
    if (coupon.length > 0) {
      setCouponLoading(true);
      setNoCoupon(false);
      api
        .post("api/coupon/", { method: "apply", title: coupon })
        .then((res) => {
          if (res.status === 200) {
            setCouponValue(res.data.value);
            setApplied(true);
            setCouponLoading(false);
          } else {
            setCouponValue(1);
            setApplied(false);
            setCouponLoading(false);
          }
        })
        .catch(() => {});
    } else {
      setNoCoupon(true);
    }
  };

  const setOrderData = (e) => {
    createOrder(true);
    if (subtotal > 1) {
      if (firstName) {
        if (address) {
          if (phone) {
            e.preventDefault();
            api
              .post("api/order/add/", {
                rest_id: restaurant,
                status: 1,
                address: address,
                coupon: coupon,
              })
              .then((res) => {
                if (res.status === 201) {
                  createOrder(false);
                  setToComplete(true);
                } else if (res.status === 303) {
                  createOrder(false);
                  setAlreadyExists(true);
                }
              })
              .catch((err) => {
                if (err.response.status === 303) {
                  createOrder(false);
                  setAlreadyExists(true);
                }
              });
            api.post("api/user/change/", { method: "name", name: firstName });
          } else {
            setNoPhone(true);
            createOrder(false);
          }
        } else {
          setNoAddress(true);
          createOrder(false);
        }
      } else {
        setNoName(true);
        createOrder(false);
      }
    } else {
      setNoItems(true);
      createOrder(false);
    }
  };

  const createAddress = (e) => {
    e.preventDefault();
    setAddressLoading(true);
    if (buildingAddress.length > 0) {
      api
        .post("api/address/add/", {
          house_address: buildingAddress,
          entrance: entrance,
          floor: floor,
          flat: flat,
        })
        .then((res) => {
          if (res.status === 201) {
            getAddressList(true);
          } else if (res.status === 409) {
            setAddressAlreadyExists(true);
            setAddressLoading(false);
          }
        })
        .catch((err) => {
          if (err.response.status === 409) {
            setAddressAlreadyExists(true);
            setAddressLoading(false);
          }
        });
    } else {
      setNoBuilding(true);
      setAddressLoading(false);
    }
  };

  const selectStyles = {
    control: (styles, { isFocused }) => ({
      ...styles,
      color: "#e8e8e8",
      backgroundColor: "#F9F9F9",
      border: isFocused ? "2px solid #6C5FBC" : "2px solid #FFF",
      fontSize: "18px",
      fontWeight: "600",
      lineHeight: "117.5%",
      transition: "0.2s",
      width: "470px",
      height: "65px",
      borderRadius: "10px",
      ":hover": {
        border: "2px solid #6C5FBC",
      },
      "::placeholder": {
        color: "#C2C3CB",
      },
    }),
    option: (styles, { isDisabled, isFocused }) => ({
      ...styles,
      backgroundColor: "#fff",
      color: isFocused ? "#6C5FBC" : "#201F1F",
      cursor: isDisabled ? "not-allowed" : "default",
      fontSize: "17px",
      fontWeight: "400",
      lineHeight: "117.5%" /* 23.5px */,
    }),
  };

  if (toComplete) {
    return <Navigate to="/complete-order" />;
  } else if (!addAddress) {
    return (
      <div id="order">
        <div id="costs">
          {applied === false && <p className="error">Invalid coupon</p>}
          {noCoupon && <p className="error">Please enter coupon to apply</p>}
          <form>
            <div id="input">
              <img src={cpImg} alt="" />
              <input
                type="text"
                placeholder="Apply Coupon"
                value={coupon}
                onChange={(e) => {
                  setCoupon(e.target.value);
                  setApplied(null);
                }}
                disabled={applied}
              />
            </div>
            <div id="coupon-btns">
              {!applied && (
                <button
                  type="submit"
                  style={{ zIndex: "10" }}
                  onClick={(e) => {
                    checkCoupon(e);
                  }}
                >
                  {couponLoading ? "Applying..." : "Apply"}
                </button>
              )}
              <button
                className="applied"
                onMouseOver={() => setDiscardCouponBtn(true)}
                style={
                  applied && !discardCouponBtn
                    ? { opacity: "1", zIndex: "5" }
                    : { opacity: "0" }
                }
              >
                ✓ Applied
              </button>
              <button
                className="remove"
                onMouseLeave={() => setDiscardCouponBtn(false)}
                style={
                  discardCouponBtn
                    ? { opacity: "1" }
                    : { opacity: "0", zIndex: "1" }
                }
                onClick={(e) => {
                  e.preventDefault();
                  setCoupon("");
                  setApplied(null);
                  setCouponValue(1);
                }}
              >
                Cancel
              </button>
            </div>
          </form>
          <h6>
            Subtotal: <span>${subtotal}</span>
          </h6>
          <h6>
            Delivery: <span>${(subtotal * 0.1).toFixed(2)}</span>
          </h6>
          <h6 style={applied ? { display: "flex" } : { display: "none" }}>
            Coupon:
            <span style={{ color: "green" }}>-${discount}</span>
          </h6>
          <h6 className="total">
            Total: <span>${getTotal()}</span>
          </h6>
        </div>
        <div id="pay">
          <div id="userInfo">
            {user.phone !== undefined ? (
              <div id="address">
                <div id="nameDiv">
                  {noName && <p className="error">Please enter your name</p>}
                  <input
                    autoComplete="off"
                    id="name"
                    type="text"
                    placeholder="How courier'll call you?"
                    value={firstName}
                    onChange={(e) => {
                      setFirstName(e.target.value);
                    }}
                  />
                </div>
                <div id="phoneDiv">
                  {noPhone && (
                    <p className="error">Please provide us phone number</p>
                  )}
                  <input
                    autoComplete="off"
                    id="name"
                    type="phone"
                    placeholder="Phone number"
                    value={phone}
                    onChange={(e) => {
                      setPhone(e.target.value);
                    }}
                  />
                </div>
              </div>
            ) : (
              <div className="input-load">Loading...</div>
            )}
            {restList.length > 0 ? (
              <div id="restDiv" data-testid="rest-select">
                <Select
                  options={restList}
                  styles={selectStyles}
                  components={{ Option: RestSelectOption }}
                  formatOptionLabel={(rest) => (
                    <div className="rest-option">
                      <img
                        src={`${import.meta.env.VITE_API_URL}/media${
                          rest.logo
                        }`}
                      />
                      <h4>{rest.label}</h4>
                    </div>
                  )}
                  placeholder="From which restaurant you want to order?"
                  onChange={changeRest}
                  isSearchable={false}
                  defaultValue={
                    restaurant
                      ? restList.find((r) => r.value === restaurant)
                      : restList[0]
                  }
                />
              </div>
            ) : (
              <div className="input-load">Loading...</div>
            )}
            {user.phone !== undefined ? (
              <div id="address" data-testid="address-select">
                {noAddress && <p className="error">Address required</p>}
                <Select
                  options={addressList}
                  styles={selectStyles}
                  placeholder="Where should we deliver your order?"
                  onChange={changeAddress}
                />
                <button id="add-address" onClick={() => setAddAddress(true)}>
                  Add address
                </button>
              </div>
            ) : (
              <div className="input-load">Loading...</div>
            )}
            {noItems && <p className="error">Your cart is empty</p>}
            {alreadyExists && (
              <p className="error">
                You already have staged order. Go{" "}
                <a href="/complete-order">here</a> to complete it
              </p>
            )}
            <button
              id="complete"
              onClick={(e) => {
                setOrderData(e);
              }}
            >
              Complete order
            </button>
          </div>
        </div>
      </div>
    );
  } else if (addAddress) {
    return (
      <div id="order">
        <div id="new-address">
          <Map
            address={buildingAddress}
            setIsLoaded={setMapLoading}
            selectMode={mapSelectMode}
            enableError={setMapsError}
            updateAddress={setBuildingAddress}
          />
          {addressAlreadyExists && (
            <p className="error" style={{ paddingBottom: "4px" }}>
              You already add this address
            </p>
          )}
          {mapsError && (
            <p className="error" style={{ paddingBottom: "4px" }}>
              {mapsError}
            </p>
          )}
          {mapLoading && (
            <div id="place-input">
              {noBuilding && (
                <p className="error" style={{ paddingBottom: "4px" }}>
                  Building address is required
                </p>
              )}
              <PlaceInput
                finalAddress={(a) => {
                  setBuildingAddress(a);
                }}
                enableSelectMode={setMapSelectMode}
                userLocationAddress={buildingAddress}
              />
            </div>
          )}
          <form>
            <div id="inputs">
              {test && (
                <input
                  type="text"
                  placeholder="Building"
                  value={buildingAddress}
                  onChange={(e) => {
                    setBuildingAddress(e.target.value);
                  }}
                />
              )}
              <div>
                <input
                  type="text"
                  placeholder="Entrance"
                  value={entrance}
                  onChange={(e) => {
                    setEntrance(e.target.value);
                  }}
                />
                <input
                  type="text"
                  placeholder="Floor"
                  value={floor}
                  onChange={(e) => {
                    setFloor(e.target.value);
                  }}
                />
                <input
                  type="text"
                  placeholder="Flat"
                  value={flat}
                  onChange={(e) => {
                    setFlat(e.target.value);
                  }}
                />
              </div>
            </div>
            <div className="btns">
              <button
                className="cancel"
                onClick={() => {
                  setBuildingAddress("");
                  setEntrance("");
                  setFloor("");
                  setFlat("");
                  setAddAddress(false);
                  setNoBuilding(false);
                  setAddressAlreadyExists(false);
                  setMapSelectMode(false);
                }}
              >
                Cancel
              </button>
              <button
                className="add"
                type="submit"
                onClick={(e) => {
                  createAddress(e);
                }}
              >
                {addressLoading ? "Creating address..." : "Add address"}
              </button>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

export default MakeOrder;
