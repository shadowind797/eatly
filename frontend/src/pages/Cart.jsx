import Header from "../components/Header.jsx";
import ItemsList from "../components/Cart/ItemsList.jsx";
import BaseHeader from "../components/BaseHeader.jsx";
import MakeOrder from "../components/Cart/MakeOrder.jsx";
import {useEffect, useState} from "react";
import api from "../api.js";

function Cart() {
    const [cartItems, setCartItems] = useState([]);
    const [total, setTotal] = useState(0);


    useEffect(() => {
        getCartItems()
    }, []);

    const getCartItems = () => {
        api
            .get("api/items/cart/")
            .then((res) => res.data)
            .then((data) => {
                setCartItems(data)
                getTotal(data)
            })
    }

    const getTotal = (cartI) => {
        api
            .post("api/items/", {data: {items: cartI, method: "for_total"}})
            .then((res) => res.data)
            .then((data) => setTotal(data))
            .catch((err) => alert(err));
    }


    if (cartItems.length > 0 && total.total > 0) {
        return (
            <div id='cart'>
                <BaseHeader />
                <Header />
                <div id="main">
                    <ItemsList items={cartItems} onChange={() => {getCartItems()}}/>
                    <MakeOrder itemsC={cartItems} subtotal={total.total}/>
                </div>
            </div>
        )
    }
}

export default Cart;
