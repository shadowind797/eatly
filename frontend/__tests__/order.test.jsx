import React from 'react';
import {render, fireEvent, waitFor, act} from '@testing-library/react';
import api from '../src/api.js';
import {MemoryRouter as Router} from "react-router-dom";
import Order from '../src/pages/Order.jsx';
import Cart from "../src/pages/Cart.jsx";

jest.mock('../src/api');

describe('Order component', () => {
    const payments = [
        {
            "id": 1,
            "number": "3245 3425 3452 3452",
            "date_to": "12/42",
            "active": false,
            "name": "gfgwsreg re gegwwr",
            "owner": 1
        }, {
            "id": 2,
            "number": "3245 3425 3452 5435",
            "date_to": "12/42",
            "active": false,
            "name": "gfgwsreg re gegwwr",
            "owner": 1
        }, {
            "id": 3,
            "number": "3245 3425 3453 5444",
            "date_to": "12/42",
            "active": false,
            "name": "gfgwsreg re gegwwr",
            "owner": 1
        }, {
            "id": 4,
            "number": "1231 2312 3312 3123",
            "date_to": "12/32",
            "active": false,
            "name": "sdgverdger ergdg",
            "owner": 1
        }, {
            "id": 5,
            "number": "4353 4534 5345 3453",
            "date_to": "12/34",
            "active": false,
            "name": "gbsdf gth th",
            "owner": 1
        }, {
            "id": 6,
            "number": "2342 3423 4243 2442",
            "date_to": "12/45",
            "active": false,
            "name": "sdgdfsfd fgvdsfgs",
            "owner": 1
        }]
    const item = [{"id": 16, "status": 1, "total": 27.49, "payment": null, "address": 2}]
    const user = [{
        "id": 1,
        "username": "shadowind",
        "email": "shadowind797@gmail.com",
        "phone": "+375666999",
        "status": 1,
        "is_banned": false,
        "ban_reason": "",
        "first_name": ""
    }]
    const addresses = [{
        id: 2,
        house_address: "Mirobod street 12, Tashkent, UZ",
        entrance: "2",
        floor: "4",
        flat: "34"
    }]

    beforeEach(() => {
        api.get.mockImplementation((url) => {
            if (url.includes('api/address')) {
                return Promise.resolve({status: 200, data: addresses});
            }
            if (url.includes('api/order/')) {
                return Promise.resolve({status: 200, data: item});
            }
            if (url.includes('api/user/')) {
                return Promise.resolve({status: 200, data: user});
            }
            if (url.includes('api/payment/')) {
                return Promise.resolve({status: 200, data: payments});
            }
            return Promise.reject(new Error('Not found'));
        });

        api.post.mockImplementation((url) => {
            if (url.includes('api/order/add/')) {
                return Promise.resolve({status: 205});
            }
            if (url.includes('api/payment/add/')) {
                return Promise.resolve({status: 201});
            }
            return Promise.reject(new Error('Not found'));
        });

        api.delete.mockImplementation((url) => {
            if (url.includes('api/items/cart/delete')) {
                return Promise.resolve({status: 202});
            }
            return Promise.reject(new Error('Not found'));
        })
    })

    it("renders without crashing", async () => {
        await act(async () => {
            render(<Router><Order/></Router>);
        })
    })

    it("matches the snapshot", async () => {
        let asFragment
        await act(async () => {
            const rendered = render(<Router><Order/></Router>);
            asFragment = rendered.asFragment
        })

        expect(asFragment()).toMatchSnapshot();
    })

    it("switches payment mode and selects payment card", async () => {
        let getByText, getByTestId, asFragment
        await act(async () => {
            const rendered = render(<Router><Order/></Router>);
            getByText = rendered.getByText
            getByTestId = rendered.getByTestId
            asFragment = rendered.asFragment
        })

        const card = getByTestId('web-card');
        const paymentSelect = getByTestId('payment-select');
        fireEvent.click(card);

        fireEvent.keyDown(paymentSelect.childNodes[1], {key: 'ArrowDown'});
        await waitFor(() => getByText('**** **** **** 3452'));
        fireEvent.click(getByText('**** **** **** 3452'));

        expect(asFragment()).toMatchSnapshot();
    })

    it("adds new user payment method", async () => {
        let getByText, getByTestId, asFragment, getByPlaceholderText, queryByTestId;

        await act(async () => {
            const rendered = render(<Router><Order/></Router>);
            getByText = rendered.getByText;
            getByTestId = rendered.getByTestId;
            getByPlaceholderText = rendered.getByPlaceholderText;
            queryByTestId = rendered.queryByTestId;
            asFragment = rendered.asFragment;
        })

        await waitFor(() => getByTestId('web-card'));

        const card = getByTestId('web-card');
        fireEvent.click(card);

        const addPaymentCardButton = getByText(/Add payment card/i);
        fireEvent.click(addPaymentCardButton);

        await waitFor(() => getByPlaceholderText('MM'));

        await act(async () => {
            fireEvent.change(getByPlaceholderText("XXXX XXXX XXXX XXXX"), {target: {value: "4334 5345 3543 9999"}});
            fireEvent.change(getByPlaceholderText("CVV/CVC"), {target: {value: "245"}});
            fireEvent.change(getByPlaceholderText("MM"), {target: {value: "04"}});
            fireEvent.change(getByPlaceholderText("YY"), {target: {value: "28"}});
            fireEvent.change(getByPlaceholderText("OWNER NAME"), {target: {value: "NONAME NONAME"}});

            fireEvent.click(getByText(/Add card/i));
        });

        await waitFor(() => {
            expect(api.post).toHaveBeenCalledWith("api/payment/add/", {
                "number": "4334 5345 3543 9999",
                "cvv": "245",
                "date_to": "04/28",
                "name": "NONAME NONAME"
            })
        })
    });
})
