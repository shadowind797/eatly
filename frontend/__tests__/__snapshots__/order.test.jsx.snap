// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    id="complete-order"
  >
    <div
      id="google-map"
    >
      <div>
        Map'll be here
      </div>
    </div>
    <div
      id="order-info"
    >
      <h1>
        Complete order
      </h1>
      <div
        id="counts"
      >
        <h4>
          Total cost: 
          <span>
            $27.49
          </span>
        </h4>
        <h4>
          Name: 
          <span />
        </h4>
        <h4>
          Payment: 
          <span>
            Cash
          </span>
        </h4>
      </div>
      <div
        id="payment"
      >
        <h3>
          Select payment method
        </h3>
        <div
          id="payment-method"
        >
          <button
            class="active"
          >
            <img
              alt=""
              src="undefined/media/img/cash.svg"
            />
          </button>
          <button
            class="[object Object]"
            data-testid="web-card"
          >
            <img
              alt=""
              src="undefined/media/img/internet.svg"
            />
          </button>
          <button
            class="[object Object]"
          >
            <img
              alt=""
              src="undefined/media/img/card.svg"
            />
          </button>
          <div
            data-testid="payment-select"
            id="card-select"
            style="display: none;"
          >
            <p
              class="error"
              style="display: none;"
            >
              Payment card is required
            </p>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-3-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
                role="log"
              />
              <div
                class=" css-1araxoe-control"
              >
                <div
                  class=" css-hlgwow"
                >
                  <div
                    class=" css-1jqq78o-placeholder"
                    id="react-select-3-placeholder"
                  >
                    Select card
                  </div>
                  <div
                    class=" css-19bb58m"
                    data-value=""
                  >
                    <input
                      aria-activedescendant=""
                      aria-autocomplete="list"
                      aria-describedby="react-select-3-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-3-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1wy0on6"
                >
                  <span
                    class=" css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <p
              id="add-payment"
            >
              Add payment card
            </p>
          </div>
        </div>
      </div>
      <p />
      <div
        id="btns"
      >
        <button
          id="cancel"
        >
          Cancel
        </button>
        <button
          id="order"
        >
          Order
        </button>
      </div>
      <p
        style="display: none;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Order component switches payment mode and selects payment card 1`] = `
<DocumentFragment>
  <div
    id="complete-order"
  >
    <div
      id="google-map"
    >
      <div>
        Map'll be here
      </div>
    </div>
    <div
      id="order-info"
    >
      <h1>
        Complete order
      </h1>
      <div
        id="counts"
      >
        <h4>
          Total cost: 
          <span>
            $27.49
          </span>
        </h4>
        <h4>
          Name: 
          <span />
        </h4>
        <h4>
          Payment: 
          <span>
            Card via Internet
          </span>
        </h4>
      </div>
      <div
        id="payment"
      >
        <h3>
          Select payment method
        </h3>
        <div
          id="payment-method"
        >
          <button
            class="[object Object]"
          >
            <img
              alt=""
              src="undefined/media/img/cash.svg"
            />
          </button>
          <button
            class="active"
            data-testid="web-card"
          >
            <img
              alt=""
              src="undefined/media/img/internet.svg"
            />
          </button>
          <button
            class="[object Object]"
          >
            <img
              alt=""
              src="undefined/media/img/card.svg"
            />
          </button>
          <div
            data-testid="payment-select"
            id="card-select"
            style="display: flex; flex-direction: column; gap: 5px;"
          >
            <p
              class="error"
              style="display: none;"
            >
              Payment card is required
            </p>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-4-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
                role="log"
              />
              <div
                class=" css-1araxoe-control"
              >
                <div
                  class=" css-hlgwow"
                >
                  <div
                    class=" css-1dimb5e-singleValue"
                  >
                    **** **** **** 3452
                  </div>
                  <div
                    class=" css-19bb58m"
                    data-value=""
                  >
                    <input
                      aria-activedescendant=""
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-4-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 0; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1wy0on6"
                >
                  <span
                    class=" css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <p
              id="add-payment"
            >
              Add payment card
            </p>
          </div>
        </div>
      </div>
      <p />
      <div
        id="btns"
      >
        <button
          id="cancel"
        >
          Cancel
        </button>
        <button
          id="order"
        >
          Order
        </button>
      </div>
      <p
        style="display: none;"
      />
    </div>
  </div>
</DocumentFragment>
`;
